moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/tesseract/bullet_utils.h

```diff
@@ -0,0 +1,1155 @@
+  void getAabb(const btTransform& t_w0, btVector3& aabbMin, btVector3& aabbMax) const override
+  {
+    m_shape->getAabb(t_w0, aabbMin, aabbMax);
+    btVector3 min1, max1;
+    m_shape->getAabb(t_w0 * m_shape_transform, min1, max1);
+    aabbMin.setMin(min1);
+    aabbMax.setMax(max1);
+  }
+
+  void getAabbSlow(const btTransform& /*t*/, btVector3& /*aabbMin*/, btVector3& /*aabbMax*/) const override
+  {
+    throw std::runtime_error("shouldn't happen");
+  }
+
+  void setLocalScaling(const btVector3& /*scaling*/) override
+  {
+  }
+
+  const btVector3& getLocalScaling() const override
+  {
+    static btVector3 out(1, 1, 1);
+    return out;
+  }
+
+  void setMargin(btScalar /*margin*/) override
+  {
+  }
+
+  btScalar getMargin() const override
+  {
+    return 0;
+  }
+
+  int getNumPreferredPenetrationDirections() const override
+  {
+    return 0;
+  }
+
+  void getPreferredPenetrationDirection(int /*index*/, btVector3& /*penetrationVector*/) const override
+  {
+    throw std::runtime_error("not implemented");
+  }
+
+  void calculateLocalInertia(btScalar, btVector3&) const override
+  {
+    throw std::runtime_error("not implemented");
+  }
+
+  const char* getName() const override
+  {
+    return "CastHull";
+  }
+
+  btVector3 localGetSupportingVertexWithoutMargin(const btVector3& v) const override
+  {
+    return localGetSupportingVertex(v);
+  }
+};
+
+/** \brief Computes the local supporting vertex of a convex shape.
+ *
+ *  If multiple vertices with equal support products exists, their average is calculated and returned.
+ *
+ *  \param shape The convex shape to check
+ *  \param localNormal The support direction to search for in shape local coordinates
+ *  \param outsupport The value of the calculated support mapping
+ *  \param outpt The computed support point */
+inline void GetAverageSupport(const btConvexShape* shape, const btVector3& localNormal, float& outsupport,
+                              btVector3& outpt)
+{
+  btVector3 ptSum(0, 0, 0);
+  float ptCount = 0;
+  float maxSupport = -1000;
+
+  const btPolyhedralConvexShape* pshape = dynamic_cast<const btPolyhedralConvexShape*>(shape);
+  if (pshape)
+  {
+    int nPts = pshape->getNumVertices();
+
+    for (int i = 0; i < nPts; ++i)
+    {
+      btVector3 pt;
+      pshape->getVertex(i, pt);
+
+      float sup = pt.dot(localNormal);
+      if (sup > maxSupport + BULLET_EPSILON)
+      {
+        ptCount = 1;
+        ptSum = pt;
+        maxSupport = sup;
+      }
+      else if (sup < maxSupport - BULLET_EPSILON)
+      {
+      }
+      else
+      {
+        ptCount += 1;
+        ptSum += pt;
+      }
+    }
+    outsupport = maxSupport;
+    outpt = ptSum / ptCount;
+  }
+  else
+  {
+    outpt = shape->localGetSupportingVertexWithoutMargin(localNormal);
+    outsupport = localNormal.dot(outpt);
+  }
+}
+
+/** @brief Check if a collision check is required between the provided collision objects
+ *  @param cow1 The first collision object
```

Comment:
why is it called cow?

Reply:
Collision Object Wrapper (COW) as it wraps around a bullet collision object `btCollisionObject`

Reply:
MoveIt style is verbose variable names, so I would call it coll_obj_wrapper or even collision_object_wrapper.

Reply:
if i could vote for a single abbreviation in the bullet code, it would be `cow` - it is used all over the code and writing `collision_object_wrapper` for each instance would lead to many ugly line breaks which in my opinion decrease readability.

Reply:
I fully understand you. IMHO, going for cow is fine. Elsewhere we used co for collision_object as well.

Reply:
I don't have a strong position, but in my last PR there were wishes to spell it out, which I did and preferred. I'd probably vote for `coll_obj_wrapper`, but maybe you can link a few spots where `cow` would be significantly better?

Reply:
We have ~ 200 occurrences of cow in the code. Replacing leads in many lines to line breaks like
![image](https://user-images.githubusercontent.com/19279526/61282836-58fa4e00-a7bc-11e9-8025-d10d6f26e91b.png)

What I could do is replace the type specifier (not variable names) with the longer `CollisionObjectWrapper`. This is not in too many places and then one can simply peek at the function header to see the type of `cow` if this abbreviation is used as a variable name.

Reply:
I think the longer type specifier is a better choice.

Reply:
I wouldn't be too concerned about line breaks. I've gotten used to Google-style code which is limited to 80 char width, *and* they use verbose variable names. I think readability of full names is better than worrying about line count.

Otherwise I keep thinking about milk and farms :cow:

Reply:
I changed all type specifiers to `CollisionObjectWrapper`